### ISSUE TITLE
Support Shopify API version 2020-07

### ISSUE THREAD

**Issue by nozzlegear** [nozzlegear/ShopifySharp#532]
Now that 2020-07 is available, and 2019-07 is the oldest supported version, I want to start updating the package to target 2020-07. 

I might be interested in exploring a different versioning scheme for the package based on the Shopify API version. Something like 2020.07.0, 2020.07.1, etc. I think we've talked about it elsewhere in this repo and I decided against it at the time, but I can't find a link at the moment. I'd be interested in hearing arguments for or against changing the versioning scheme. For reference, we're currently using the typical SemVer scheme e.g. `5.0.0`, `5.0.1`, `5.1.0`, etc.

## TODO

- [x] Implement the new Collections endpoint (smart collections no longer supported on current Collects endpoint) (#539)

**Comment by drovani** [comment]
I think a combination of SemVer and ShopifyAPI matching would be appropriate. Since there are frequently non-breaking changes between Shopify API versions, there is usually very little work with respect to keeping ShopifySharp up to date. The 2019-04 to 2019-07 jump was really the most recent painful upgrade, right?

I could see doing `5.202007.0`, `5.202007.1`, etc. for releases between Shopify API versions. If 2020-10 has no breaking changes, then `5.202010.0`, `5.202010.1`. However, if 2021-01 has some breaking changes, then bump the major version to `6.202101.1`. Maybe I'm overthinking it and this just makes everything more confusing. If so, then I think it makes most sense to stick with the Shopify API version, and completely leave it up to the consuming developer to figure out what has changed and what hasn't.

If there are bug fixes in `2020.07.02` would you back-port them to `2020.04.*`?

**Comment by clement911** [comment]
Found the original discussion: https://github.com/nozzlegear/ShopifySharp/pull/355#issuecomment-495422024

**Comment by clement911** [comment]
Luckily upgrading to more recent versions will be MUCH MUCH easier that the previous time.
I had a quick look at the [release notes](https://shopify.dev/concepts/about-apis/versioning/release-notes) and these are the changes I have found (related to the admin REST API only).

2020-01
**Breaking change**: Smart collections cannot use the Collect endpoint anymore => Must implement new Collections endpoint.
https://shopify.dev/docs/admin-api/rest/reference/products/collection

New Fulfillment orders endpoint: would be nice but not necessarily required since it is a new service so there would be no breaking change

2020-04
New fields on Location, Order, Refund, AbandonedCheckout, Fulfillment

2020-07
ProductVariant.images is removed

That's not much.
The biggest piece will be implementing the new Collection endpoint in 2020-01.
As a result, I think it's perfectly reasonable to upgrade ShopifySharp straight to the latest 2020-07 version.

Regarding ShopifySharp's versioning, I have gone back and forth but my latest personal preference would be to keep doing things as we have done so far. That is, normal semver without incorporating the underlying Shopify API version (e.g. 202001) into it. Incorporating the API version into the lib version will likely break the semver semantics (used by NuGet) and / or make it awkward. 

 I think that documenting the API version used in the `readme.md` is enough.

**Comment by nozzlegear** [comment]
I agree, switching away from semver to follow Shopify's versioning doesn't make sense for this project and for NuGet. Adding an API version -> ShopifySharp version table to the readme should be enough!

**Comment by clement911** [comment]
@nozzlegear can you please create a branch for the 2020-07 release?
I'd like to start contributing.

**Comment by nozzlegear** [comment]
Done! You should be able to make prs to the 2020-07 branch now.

--
Josh Harms

**Comment by clement911** [comment]
Thanks!

**Comment by clement911** [comment]
@nozzlegear thanks for implementing the new CollectionService.

Now that it's done, is there any reason not to pin the lib to 2020-07?
A few fields of the new fields are not there, but so what?

Also, I find it strange that you pinned the CollectionService to 2020-07 by override the lib version.
I would prefer that the entire lib is pinned to a single version, as it was before. Otherwise, it makes it quite hard to reason about.

Actually, I think that [ShopifyService.APIVersion](https://github.com/nozzlegear/ShopifySharp/blob/master/ShopifySharp/Services/ShopifyService.cs#L18) should be a `public const string` instead of `public virtual string`. I believe that's how Stripe does it. One lib = One API version.

What do you think?

**Comment by clement911** [comment]
It might be warranted to increase the major version and have clear release notes to explain that the existing CollectService cannot be used anymore with smart collections..

**Comment by rohanp1903** [comment]
Hi @nozzlegear , it seems that we can easily create or update products with SEO title and description metafields using the new API version. Hope you will integrate this while upgrading to 2020-07 version.

[https://shopify.dev/docs/admin-api/rest/reference/products/product?api[version]=2020-07](https://shopify.dev/docs/admin-api/rest/reference/products/product?api[version]=2020-07)

Update a product's SEO title and description
PUT /admin/api/2020-07/products/632910392.json
{
  "product": {
    "id": 632910392,
    "metafields_global_title_tag": "Brand new title",
    "metafields_global_description_tag": "Brand new description"
  }
}

**Comment by clement911** [comment]
@nozzlegear I would love to hear your comments on my latest replies above.
We are approaching the October 1st deadline fast so it would be great to get things in order and get on the same page.

**Comment by nozzlegear** [comment]
Sure! I'd like to leave it as `public virtual string` so that devs can manually override the version if they want to implement their own custom service. I don't mind bumping to 6.0 due to the collections change, but I'm not sure it warrants it. Maybe I'm mistaken, but the service still works with normal collections and doesn't _specifically_ use anything that is smart collection only. If we bumped to 6.0, are there any methods or objects that would change or be removed?

Regarding using a different API version for the new Collections service, I think that's going to be an exception and hopefully not something we need to do often. I do want to keep the option around though for time sensitive cases like this one, where we may not be ready to release a whole migration to a new API version. I agree that if each service starts targeting random API versions it will be very hard to reason about the package and the documentation. I absolutely want to improve the documentation for the package and specifically document which API version is targeted/supported, and I'm wondering if there's a way to make the intellisense for each service include that information dynamically (along with maybe a link to Shopify's documentation for the endpoint). 

Sorry, kind of rambling here, I'm up a little too late tonight! Happy to discuss this more.

**Comment by clement911** [comment]
I've submitted a PR https://github.com/nozzlegear/ShopifySharp/pull/568 to merge into the 2020-07.

I believe it's even safe to merge 2020-07 into master and release.

**Comment by nozzlegear** [comment]
Alright, #568 has been merged and published! I've been thinking about the decision to increment to 6.0 or stay in 5.0 branch. I came really close to incrementing to 6.0, but ultimately I decided to keep us in 5.0 after I realized I wouldn't need to write a migration guide for 5.* => 6.0 because nothing major changed. 

I understand Shopify's change to `CollectService.ListAsync` is going start returning errors when passing in a smart collection ID, but in the end the build itself doesn't break and no signature has changed, so I feel like incrementing to 6.0 is jumping the gun. If a dev using this package were to upgrade from 5.5 to 6.0, they still wouldn't necessarily know that they can't pass in smart collection IDs to that method until it throws an error. 

We can certainly make it easier for devs to know which version of Shopify's API they're using though, I definitely agree it's confusing and not easy to figure out. To help alleviate that, I've added [an API support table to the readme](https://github.com/nozzlegear/ShopifySharp#api-support). I'd also like to look into adding the API version to intellisense somehow. 

Regarding the issue itself, ShopifySharp now targets API version 2020-07 in v5.6.0, now published on Nuget. Thank you @clement911 in particular for the help as always! Much appreciated!
